File: go/14_reasoning.go
package main

/*
Example 13: Reasoning with openai/gpt-oss-20b

Demonstrates:
- Reasoning models that show step-by-step thinking
- Three reasoning formats: raw, parsed, hidden
- Three reasoning effort levels: low, medium, high
- Cache optimization for cost savings

What you'll learn:
- How reasoning models differ from chat models
- When to use different reasoning formats
- How to optimize message order for cache hits
- Token usage and cost tracking

Note: Uses Go standard library only
*/

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

func callReasoning(messages []interface{}, formatType *string, effort string) (map[string]interface{}, error) {
	requestData := map[string]interface{}{
		"model":                 "openai/gpt-oss-20b",
		"messages":              messages,
		"temperature":           0.6,
		"max_completion_tokens": 1024,
		"reasoning_effort":      effort,
	}

	if formatType != nil {
		requestData["reasoning_format"] = *formatType
	}

	requestBody, _ := json.Marshal(requestData)

	req, err := http.NewRequest("POST", "https://api.demeterics.com/groq/v1/chat/completions", bytes.NewBuffer(requestBody))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", os.Getenv("DEMETERICS_API_KEY")))

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	var result map[string]interface{}
	json.Unmarshal(body, &result)
	return result, nil
}

func main() {
	// Check for API key
	if os.Getenv("DEMETERICS_API_KEY") == "" {
		fmt.Fprintln(os.Stderr, "Error: DEMETERICS_API_KEY not set")
		fmt.Fprintln(os.Stderr, "Get your Managed LLM Key from: https://demeterics.com")
		os.Exit(1)
	}

	fmt.Println("========================================")
	fmt.Println("Reasoning with openai/gpt-oss-20b")
	fmt.Println("========================================\n")

	query := "How many 'r' letters are in the word 'strawberry'? Think through this step-by-step."
	messages := []interface{}{
		map[string]interface{}{"role": "user", "content": query},
	}

	fmt.Printf("Query: %s\n\n", query)

	// Demo 1: Raw format (default)
	fmt.Println("========================================")
	fmt.Println("Demo 1: Raw Format (reasoning in <think> tags)")
	fmt.Println("========================================\n")

	rawResponse, err := callReasoning(messages, nil, "medium")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	choices := rawResponse["choices"].([]interface{})
	message := choices[0].(map[string]interface{})["message"].(map[string]interface{})
	content := message["content"].(string)

	fmt.Println("Response (raw format):")
	fmt.Println(content)
	fmt.Println()

	usage := rawResponse["usage"].(map[string]interface{})
	promptTokens := usage["prompt_tokens"].(float64)
	completionTokens := usage["completion_tokens"].(float64)
	totalTokens := usage["total_tokens"].(float64)

	fmt.Println("Token Usage (Demo 1):")
	fmt.Printf("  Prompt tokens: %.0f\n", promptTokens)
	fmt.Printf("  Completion tokens: %.0f\n", completionTokens)
	fmt.Printf("  Total tokens: %.0f\n\n", totalTokens)

	// Demo 2: Parsed format
	fmt.Println("========================================")
	fmt.Println("Demo 2: Parsed Format (separate reasoning field)")
	fmt.Println("========================================\n")

	parsedFormat := "parsed"
	parsedResponse, _ := callReasoning(messages, &parsedFormat, "medium")

	parsedChoices := parsedResponse["choices"].([]interface{})
	parsedMessage := parsedChoices[0].(map[string]interface{})["message"].(map[string]interface{})

	if reasoning, ok := parsedMessage["reasoning"].(string); ok {
		fmt.Println("Reasoning process:")
		fmt.Println(reasoning)
		fmt.Println()
	}

	fmt.Println("Final answer:")
	fmt.Println(parsedMessage["content"].(string))
	fmt.Println()

	// Demo 3: Hidden format
	fmt.Println("========================================")
	fmt.Println("Demo 3: Hidden Format (only final answer)")
	fmt.Println("========================================\n")

	hiddenFormat := "hidden"
	hiddenResponse, _ := callReasoning(messages, &hiddenFormat, "medium")

	hiddenChoices := hiddenResponse["choices"].([]interface{})
	hiddenMessage := hiddenChoices[0].(map[string]interface{})["message"].(map[string]interface{})

	fmt.Println("Response (hidden format):")
	fmt.Println(hiddenMessage["content"].(string))
	fmt.Println()

	// Cost calculation
	fmt.Println("========================================")
	fmt.Println("Cost Analysis")
	fmt.Println("========================================\n")

	inputCostPer1M := 0.10
	cachedInputCostPer1M := 0.05
	outputCostPer1M := 0.50

	inputCost := (promptTokens * inputCostPer1M) / 1000000
	outputCost := (completionTokens * outputCostPer1M) / 1000000
	totalCost := inputCost + outputCost

	fmt.Println("Pricing (openai/gpt-oss-20b):")
	fmt.Println("  Input tokens: $0.10 per 1M")
	fmt.Println("  Cached input tokens: $0.05 per 1M (50% discount!)")
	fmt.Println("  Output tokens: $0.50 per 1M\n")

	fmt.Println("This request:")
	fmt.Printf("  Input cost: $%.6f (%.0f tokens)\n", inputCost, promptTokens)
	fmt.Printf("  Output cost: $%.6f (%.0f tokens)\n", outputCost, completionTokens)
	fmt.Printf("  Total cost: $%.6f\n\n", totalCost)

	// Cache optimization explanation
	fmt.Println("========================================")
	fmt.Println("Cache Optimization Strategy")
	fmt.Println("========================================\n")

	fmt.Println("What is cached input?")
	fmt.Println("Groq caches parts of your prompt to speed up responses and reduce costs.")
	fmt.Println("Cached tokens are 50% cheaper ($0.05/1M vs $0.10/1M for this model).\n")

	fmt.Println("How to maximize cache hits:")
	fmt.Println("Order messages from most constant to most variable:\n")
	fmt.Println("  1. System prompts (always the same across requests)")
	fmt.Println("  2. Context/documents (same during a conversation)")
	fmt.Println("  3. User questions (changes with each request)\n")

	fmt.Println("Example: agentresume.ai recruiter chatbot\n")
	fmt.Println("Optimal message order:")
	fmt.Println(`[
  {
    "role": "system",
    "content": "You are a recruiter assistant. Analyze resumes..."
  },                                                    ← Always cached
  {
    "role": "user",
    "content": "Candidate Resume:\nJohn Doe\nSkills: Python, Go..."
  },                                                    ← Cached per conversation
  {
    "role": "user",
    "content": "Does this candidate have Python experience?"
  }                                                     ← Not cached (changes each Q)
]
`)

	fmt.Println("Cache benefit calculation:")
	fmt.Println("If 1000 tokens are always cached (system + resume):")
	fmt.Println("  Normal cost: 1000 × $0.10 / 1M = $0.0001")
	fmt.Println("  Cached cost: 1000 × $0.05 / 1M = $0.00005")
	fmt.Println("  Savings: 50% per request!\n")

	fmt.Println("========================================")
	fmt.Println("Reasoning Effort Comparison")
	fmt.Println("========================================\n")

	fmt.Println("Reasoning effort levels:")
	fmt.Println("  low    - Quick analysis, fewer reasoning tokens")
	fmt.Println("  medium - Balanced reasoning (default)")
	fmt.Println("  high   - Deep analysis, more reasoning tokens\n")
	fmt.Println("Use 'low' for simple questions, 'high' for complex reasoning tasks.\n")

	fmt.Println("========================================")
	fmt.Println("Summary")
	fmt.Println("========================================\n")
	fmt.Println("✓ Demonstrated three reasoning formats")
	fmt.Println("✓ Showed token usage and costs")
	fmt.Println("✓ Explained cache optimization strategy")
	fmt.Println("✓ Model excels at step-by-step logical reasoning\n")
}

File: cpp/14_reasoning.cpp
/**
 * ============================================================================
 * Example 14: Reasoning - Step-by-Step Thinking with Prompt Caching
 * ============================================================================
 *
 * WHAT THIS DEMONSTRATES:
 *   - Using AI for complex reasoning tasks
 *   - Chain-of-thought prompting
 *   - Prompt caching for cost savings
 *
 * WHAT YOU'LL LEARN:
 *   - How to get AI to "think" through problems
 *   - Structured reasoning prompts
 *   - Caching benefits for repeated queries
 *
 * COMPILE:
 *   g++ -std=c++17 -o 14_reasoning 14_reasoning.cpp -lcurl -I lib/
 *
 * RUN:
 *   ./14_reasoning
 *
 * ============================================================================
 */

#include <iostream>
#include <string>
#include <cstdlib>
#include <curl/curl.h>
#include "json.hpp"

using json = nlohmann::json;

static size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    std::string *response = static_cast<std::string*>(userp);
    response->append(static_cast<char*>(contents), size * nmemb);
    return size * nmemb;
}

int main() {
    /* Step 1: Get API key */
    const char *api_key_cstr = std::getenv("DEMETERICS_API_KEY");
    if (api_key_cstr == nullptr) {
        std::cerr << "Error: DEMETERICS_API_KEY not set\n";
        return 1;
    }
    std::string api_key(api_key_cstr);

    /* Step 2: Initialize curl */
    curl_global_init(CURL_GLOBAL_ALL);
    CURL *curl = curl_easy_init();
    if (curl == nullptr) {
        curl_global_cleanup();
        return 1;
    }

    /* Step 3: Define a problem requiring reasoning */
    std::string reasoning_problem = R"(A farmer has 17 sheep. All but 9 run away. How many sheep does the farmer have left?

Think through this step-by-step before giving your final answer.)";

    std::cout << "=== AI REASONING ===\n";
    std::cout << "Using chain-of-thought prompting for complex problems\n\n";
    std::cout << "Problem: " << reasoning_problem << "\n\n";

    /* Step 4: Build request with reasoning system prompt */
    std::string system_prompt = R"(You are a logical reasoning assistant. When solving problems:
1. Read the problem carefully - note any tricky wording
2. Identify what is being asked
3. Work through the logic step by step
4. Check your reasoning for errors
5. Provide a clear final answer

Take your time and think carefully before answering.)";

    json request_body = {
        {"model", "openai/gpt-oss-20b"},  /* Good for reasoning */
        {"messages", {
            {{"role", "system"}, {"content", system_prompt}},
            {{"role", "user"}, {"content", reasoning_problem}}
        }},
        {"max_tokens", 500},
        {"temperature", 0.3}  /* Lower = more focused */
    };

    std::string json_payload = request_body.dump();

    /* Step 5: Set up headers */
    struct curl_slist *headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/json");
    std::string auth_header = "Authorization: Bearer " + api_key;
    headers = curl_slist_append(headers, auth_header.c_str());

    /* Step 6: Configure and perform request */
    std::string response_data;
    curl_easy_setopt(curl, CURLOPT_URL, "https://api.demeterics.com/groq/v1/chat/completions");
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_payload.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_data);

    std::cout << "AI is thinking...\n\n";
    CURLcode res = curl_easy_perform(curl);

    if (res != CURLE_OK) {
        std::cerr << "Error: " << curl_easy_strerror(res) << "\n";
    } else {
        /* Step 7: Parse and display results */
        try {
            json response = json::parse(response_data);

            if (response.contains("choices") && !response["choices"].empty()) {
                std::string answer = response["choices"][0]["message"]["content"];
                std::cout << "=== AI REASONING CHAIN ===\n\n";
                std::cout << answer << "\n";
            }

            /* Token usage - check for caching */
            if (response.contains("usage")) {
                auto usage = response["usage"];
                std::cout << "\n=== TOKEN USAGE ===\n";
                std::cout << "Prompt tokens: " << usage["prompt_tokens"] << "\n";
                std::cout << "Completion tokens: " << usage["completion_tokens"] << "\n";

                if (usage.contains("prompt_cache_hit_tokens") &&
                    usage["prompt_cache_hit_tokens"].get<int>() > 0) {
                    std::cout << "Cached tokens: " << usage["prompt_cache_hit_tokens"]
                              << " (50% discount!)\n";
                }
            }

        } catch (const json::exception &e) {
            std::cerr << "Parse error: " << e.what() << "\n";
            std::cout << "Raw response: " << response_data << "\n";
        }
    }

    /* Step 8: Clean up */
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
    curl_global_cleanup();

    /* Educational info */
    std::cout << "\n=== REASONING TECHNIQUES ===\n\n";
    std::cout << "CHAIN-OF-THOUGHT PROMPTING:\n";
    std::cout << "  Ask the AI to \"think step by step\"\n";
    std::cout << "  Helps with math, logic, and complex problems\n\n";

    std::cout << "PROMPT CACHING:\n";
    std::cout << "  - System prompts can be cached\n";
    std::cout << "  - Cached tokens cost 50% less\n";
    std::cout << "  - Great for repeated similar queries\n\n";

    std::cout << "TIPS FOR BETTER REASONING:\n";
    std::cout << "  1. Use lower temperature (0.1-0.3)\n";
    std::cout << "  2. Ask to \"check your work\"\n";
    std::cout << "  3. Break complex problems into parts\n";
    std::cout << "  4. Use explicit step-by-step instructions\n";

    return 0;
}

/*
 * C++ RAW STRING LITERALS:
 *
 * R"(...)" allows multi-line strings without escaping.
 * Perfect for long prompts and JSON templates!
 *
 * REASONING NOTES:
 *
 * CHAIN-OF-THOUGHT:
 *   Just adding "think step by step" can dramatically
 *   improve AI performance on reasoning tasks.
 *
 * PROMPT CACHING:
 *   - System prompts repeated across requests can be cached
 *   - Reduces latency and cost
 *   - Especially useful for long system prompts
 */
